src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
-    XYShape.T.fromArrays(xs, ys) |> Distributions.Continuous.make(`Linear, _);
+  let combinationDistributionOfXYShapes = (sc1: DistTypes.continuousShape, // continuous shape
+                                        sd1: DistTypes.discreteShape, // discrete shape
+                                        sc2: DistTypes.continuousShape,
+                                        sd2: DistTypes.discreteShape, func): (DistTypes.continuousShape, DistTypes.discreteShape) => {
+
+    // First, deal with the discrete-discrete convolution:
+    let (ddxs, ddys) = jsDiscreteCombinationConvolve(sd1.xs, sd1.ys, sd2.xs, sd2.ys, func);
+    let ddxy: DistTypes.discreteShape = {xs: ddxs, ys: ddys};
+
+    // Then, do the other three:
+    let downsample = (sc: DistTypes.continuousShape) => {
+      let scLength = E.A.length(sc.xyShape.xs);
+      let scSqLength = sqrt(float_of_int(scLength));
+      scSqLength > 10. ? Distributions.Continuous.T.truncate(int_of_float(scSqLength), sc) : sc;
+    };
+
+    let combinePointConvolutionResults = ca => ca |> E.A.fmap(s => {
+        // s is an array of (x, y) objects
+        let (xs, ys) = Belt.Array.unzip(s);
+        Distributions.Continuous.make(`Linear, {xs, ys});
+      })
+    |> Distributions.Continuous.reduce((+.));
+
+    let sc1d = downsample(sc1);
+    let sc2d = downsample(sc2);
+
+    let ccxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let dcxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let cdxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let continuousShapeSum = Distributions.Continuous.reduce((+.), [|ccxy, dcxy, cdxy|]);
+
+    (continuousShapeSum, ddxy)
   };
 
-  let toShape = (dists: t, sampleCount: int) => {
-    let normalized = normalizeWeights(dists);
-    let continuous =
-      normalized
-      |> E.A.filter(((r, _)) => GenericSimple.contType(r) == `Continuous)
-      |> continuousShape(_, sampleCount);
-    let discrete =
-      normalized
-      |> E.A.filter(((r, _)) => GenericSimple.contType(r) == `Discrete)
-      |> discreteShape(_, sampleCount);
-    let shape =
-      MixedShapeBuilder.buildSimple(~continuous=Some(continuous), ~discrete);
-    shape |> E.O.toExt("");
+  let evaluateCombinationDistribution = (et1: nodeResult, et2: nodeResult, op: operation, n: int) => {
+    /* return either a Distribution or a RenderedShape. Must integrate to 1. */
+
+    let func = funcFromOp(op);
+    switch ((et1, et2, op)) {
+    /* Known cases: replace symbolic with symbolic distribution */
+    | (`Simple(`Float(v1)), `Simple(`Float(v2)), _) => {
+        `Simple(`Float(func(v1, v2)))
+      }
+
+    | (`Simple(`Normal(n2)), `Simple(`Float(v1)), `AddOperation)
```

Comment:
It's cool to me that you have all these special cases to handle these correctly. However, I'm curious if there can be a cleaner way to handle these. Right now the code of "how to analytically to operations for different distributions" is pretty interspersed with all of the other logic.

One way around this could be something like we have multiple stages of going through the tree. Maybe in the first stage, we just check for things like this, and deal with all of them. This is similar to having a compiler that works in discrete steps, rather than trying to do everything at once.

Then we could do:
```
distTree |> applyAnalyticalSimplifications |> toRenderedDistribution
```
Or similar.

This doesn't have to come now though; maybe in a future refactor.

Reply:
I'd also recommend adding functionality to the ``Normal`` module and then referencing it, rather than adding the operations directly in places like this.